### ISSUE TITLE
[Atom 1.32.0] Does not inherit environment variables when launched not from the command line

### ISSUE THREAD

**Issue by jzazo** [atom/atom#18318]
### Prerequisites

* [X] Put an X between the brackets on this line if you have done all of the following:
    * Reproduced the problem in Safe Mode: https://flight-manual.atom.io/hacking-atom/sections/debugging/#using-safe-mode
    * Followed all applicable steps in the debugging guide: https://flight-manual.atom.io/hacking-atom/sections/debugging/
    * Checked the FAQs on the message board for common solutions: https://discuss.atom.io/c/faq
    * Checked that your issue isn't already filed: https://github.com/issues?utf8=✓&q=is%3Aissue+user%3Aatom
    * Checked that there is not already an Atom package that provides the described functionality: https://atom.io/packages

### Description

When I launch Atom in Ubuntu 18.04 it does not inherit environmental variables. Previous versions did inherit them. My suspicion is that since the startup script has been reduced, a regression might have been introduced.

If I launch Atom from command line, it does inherit variables correctly.

*Edit by @rsese to add explicit repro steps based on https://github.com/atom/atom/issues/18318#issuecomment-432821319, expected/actual behavior, and how often it reproduces*

### Steps to Reproduce 

1. Edit and save `~/.bashrc` to change `$PATH`, added this at the end:

```bash
export PATH="$PATH:my/new/path"
```

2. Start Atom from Activities or your Favorites
3. Open devtools <kbd>ctrl-shift-i</kbd>
4. In Console, run `process.env["PATH"]`

**Expected behavior:** [What you expect to happen]

Change in step 1 to be available.

**Actual behavior:** [What actually happens]

Change in step 1 is not available.

**Reproduces how often:** [What percentage of the time does it reproduce?]

100% with repro steps and if you start Atom from Activities or Favorites, doesn't reproduce if you start Atom from the command line.  

Doesn't reproduce on Atom 1.31.1.

### Versions

```
Atom    : 1.32.0
Electron: 2.0.9
Chrome  : 61.0.3163.100
Node    : 8.9.3
```

*Edit by @rsese to mention the OS confirmed on*

Ubuntu 18.04.1 (Also noticed that this doesn't reproduce on macOS if I start Atom from spotlight or the dock.)

**Comment by rsese** [comment]
Thanks for the report!

Just to clarify, can you share specifically how you're seeing this?  Are you opening devtools console and checking for a particular variable in `process.env`?  Also, can you add specific steps to reproduce to the issue body (I tried starting Atom from Activities and from the favorites icon I have pinned but couldn't seem to reproduce)?

**Comment by jzazo** [comment]
Hi! Thanks for the response. I noticed the problem when several packages started to fail, such as flake8, black, or julia. These packages look for python/julia paths  (anaconda3 for the first two packages), which I export in my .bashrc file. If I initialize Atom from terminal, all packages work well again.

Errors I get from these packages:
`Error: Failed to spawn command `flake8`. Make sure `flake8` is installed and on your PATH`
`We tried to launch Julia from: juliaThis path can be changed in the settings. Details:   /bin/sh: 1: julia: not found`
`Black failed to reformat 💥 💔 💥 black: not found`

I checked `process.env` as you suggest. Anaconda and julia paths are not exported in the `PATH` variable if I initialize Atom from the dash. When I initialize from terminal, they do appear, consistent with previous behavior.

**Comment by rsese** [comment]
> To be more explicit. In my `.bashrc` file I export the anaconda path, e.g.
> `export PATH="/opt/anaconda3/bin:$PATH"`
> This is not seen by Atom in `process.env` if I initialize from favourites, or activities.

Ahh thanks @jzazo :+1:  I was able to reproduce with these steps on 1.32.0.  I've updated the issue body with these steps:

1. Edit and save `~/.bashrc` to change `$PATH`, added this at the end:

```bash
export PATH="$PATH:my/new/path"
```

2. Start Atom from Activities or your Favorites
3. Open devtools <kbd>ctrl-shift-i</kbd>
4. In Console, run `process.env["PATH"]`

Expect to see my change in step 1, actual is I don't see my changes in step 1.

Also confirmed that I see your expected behavior in 1.31.1 and that if I start Atom from the command line in 1.32.0, the changes in step 1 are available.

Also noticed that this doesn't reproduce on macOS if I start Atom from spotlight or the dock.

**Comment by maxbrunsfeld** [comment]
So far, I can't reproduce this on my Ubuntu 16.04 VM.

I'm opening Atom from the desktop with this thing (not sure what it's called):

![screen shot 2018-11-15 at 12 04 08 pm](https://user-images.githubusercontent.com/326587/48578486-983f6a80-e8ce-11e8-88a5-c18fe5a268dd.png)

My `.bashrc` contains:

```sh
export PATH=$PATH:/a/new/path
```

And I see the new exported path when I evaluate `process.env.PATH` in the dev tools.

**Comment by maxbrunsfeld** [comment]
I don't know very much about the way linux applications work, but I think https://github.com/atom/atom/pull/17508 may have caused this problem.

That PR changed the `Exec` field of the `.desktop` file to point to the Atom *shell* script instead of the Atom binary. My question is - what kind of shell does Linux run when you launch a shell script from a desktop shortcut? Maybe the shell does not load your `.bashrc` file?

Previously, we were able to detect that Atom was not launched from a shell at all by the absence of a `PWD` environment variable. When we detected that condition, we would manually spawn a login shell in order to find out what environment variables were exported in `.bashrc`.

/cc @rayfoss since you created that PR. I don't have context on why you made that change and what root problem it fixed. Could you provide some insight?

**Comment by maxbrunsfeld** [comment]
@jzazo One additional question - if you export environment variables in the `.profile` or `.login` script instead of `.bashrc`, do you still see this problem?

**Comment by jzazo** [comment]
I think you are right. What I export in .profile is loaded by Atom, but not .bashrc.

I found the following paragraph in this [link](https://help.ubuntu.com/community/EnvironmentVariables):

> Shell config files such as ~/.bashrc, ~/.bash_profile, and ~/.bash_login are often suggested for setting environment variables. While this may work on Bash shells for programs started from the shell, variables set in those files are not available by default to programs started from the graphical environment in a desktop session.

**Comment by maxbrunsfeld** [comment]
Ok, so the cleanest workaround right now would be to add this to your `.profile`:

```sh
source ~/.bashrc
```

That doesn't mean this isn't a bug though. But overall, I'm a little confused as to what our expected behavior *should* be, and how that was changed by https://github.com/atom/atom/pull/17508. I'm going to leave this open to allow other linux users to comment.

**Comment by jzazo** [comment]
I tried your suggestion, but it didn't work. I tried both `source ~/.bashrc` and `source $HOME/.bashrc`, but the environments were not imported. (I rebooted my computer after changing `~/.profile`).

More explicitly, `PATH="$PATH:/opt/anaconda3/bin/"` works if changed in `.profile` and the path is visible within Atom when loaded with desktop shortcut; but `source ~/.bashrc` does not work. My guess is that `source` is a bash command, and `.profile` is not loaded with bash for graphical applications. Any other idea? I could of course move my environments from `.bashrc` to `.profile` but I am not sure that is ideal.

**Comment by maxbrunsfeld** [comment]
> My guess is that `source` is a bash command, and .profile is not loaded with bash for graphical applications.

Oh, good point, maybe it's using `sh`. In that case, instead of `source`, you can use the `.` builtin:

```sh
. $HOME/.bashrc
```

**Comment by jzazo** [comment]
No, it didn't work either. I tried `. $HOME/.bashrc` and `$HOME/.bashrc`, as well as the hardcoded versions, and it didn't work in any case :S

**Comment by jzazo** [comment]
The reason sourcing `.bashrc` was not working is because the default template has the following lines
```
# If not running interactively, don't do anything
case $- in
    *i*) ;;
      *) return;;
esac
``` 
So if you source `.bashrc` from a non-interactive shell, it doesn't do anything. It seems that the graphical shell runs non-interactively and, therefore, it did't work. I commented those lines and now it works fine. I didn't need to change anything in `.profile`, because it already sources `.bashrc` by default:
```
# if running bash
if [ -n "$BASH_VERSION" ]; then
    # include .bashrc if it exists
    if [ -f "$HOME/.bashrc" ]; then
	. "$HOME/.bashrc"
    fi
fi
```
I found the solution [here](https://stackoverflow.com/questions/43659084/source-bashrc-in-a-script-not-working).

**Comment by jzazo** [comment]
Changing exec line of `\usr\share\applications\atom.desktop` to
```
Exec=bash -i /usr/bin/atom %F
```
seems to fix the issue. I don't know if this is the best way to load the script in interactive way. Maybe the execution call of Atom inside `\usr\bin\atom` can be changed to load interactively, alternatively. Any ideas?

**Comment by someburner** [comment]
@jzazo - thanks for the tip - that fixed my process.env issue on ubuntu 18.10. I can also confirm my $HOME/.bashrc and $HOME/.profile settings do not get picked up in atom. This is using 1.34.0 installed with `sudo dpkg -i`. but has been an issue for a long long time.

**Comment by rafeca** [comment]
Hey! 👋 First of all sorry for the delay handling this issue...

I've done some investigation and indeed this issue was introduced by #17508: Atom [has some logic](https://github.com/atom/atom/blob/master/src/update-process-env.js#L61-L95) to get the correct env variables when calling directly the Atom binary (the one located in `/usr/share/atom/atom`).

This logic is bypassed when calling Atom via the bash CLI (the script located in `/usr/bin/atom`). This is done [by checking](https://github.com/atom/atom/blob/master/src/update-process-env.js#L54) the `ATOM_DISABLE_SHELLING_OUT_FOR_ENVIRONMENT` env var which [is set](https://github.com/atom/atom/blob/master/atom.sh#L27) by the CLI (there's more info in [the PR](https://github.com/atom/atom/pull/12562) that introduced this).

This logic assumes that entry points that don't contain the correct env vars (like a desktop launcher) will call the binary directly, while a shell script or terminal (which contain the correct env vars) will call the CLI. After #17508, this is no longer true, and this is the root cause of the issue.

IMHO the best solution here is to keep using the previous logic to calculate env vars on Atom instead of setting the bash script as interactive (which can cause unexpected issues depending on the user configuration). To do so, we can either create a second CLI for desktop launcher or pass an special env var to the current CLI that avoids setting the `ATOM_DISABLE_SHELLING_OUT_FOR_ENVIRONMENT` var.

Since the second option is simpler, I'm going to implement it and send a PR with the fix.

**Comment by rafeca** [comment]
This should have got fixed by #19138, thanks for reporting! 😄
